File: packages/typescript/goldenpipe/src/core/adapters/check.ts
/**
 * GoldenCheck adapter — wraps GoldenCheck-JS `scanData`.
 * Port of goldenpipe/adapters/check.py.
 *
 * Shape divergence vs Python: the Python adapter calls `scan_file(path)` and
 * reads `ctx.metadata["source"]`. GoldenCheck-JS's edge-safe `scanData` instead
 * operates on a `TabularData` built from rows, so the TS adapter scans
 * `ctx.df` directly. This means `goldencheck.scan` succeeds in the in-memory
 * (`runDf`) path here, whereas the Python `run_df` path fails the scan stage
 * (it has no file). Use `run(source)` for cross-language parity.
 *
 * Edge-safe: no `node:` imports (GoldenCheck-JS core is edge-safe).
 */

import { scanData, TabularData, severityLabel } from "goldencheck/core";
import type { Finding, ColumnProfile } from "goldencheck/core";
import type { PipeContext, Stage, StageResult } from "../models.js";
import { StageStatus } from "../models.js";
import {
  buildContextsFromCheck,
  type ColumnProfileLike,
  type FindingLike,
} from "../columnContext.js";

interface NormalizedFinding {
  severity: string;
  check: string;
  column: string;
  message: string;
}

/** Normalize a GoldenCheck-JS Finding (numeric severity) to the dict shape
 *  the Python pipeline used (string severity label). */
function normalizeFinding(f: Finding): NormalizedFinding {
  return {
    severity: severityLabel(f.severity).toLowerCase(),
    check: f.check,
    column: f.column,
    message: f.message,
  };
}

/** Map a GoldenCheck-JS ColumnProfile to the minimal shape columnContext consumes. */
function toColumnProfileLike(cp: ColumnProfile): ColumnProfileLike {
  return {
    name: cp.name,
    inferredType: cp.inferredType,
    nullPct: cp.nullPct,
    uniqueCount: cp.uniqueCount,
  };
}

export const ScanStage: Stage = {
  info: { name: "goldencheck.scan", produces: ["findings", "profile"], consumes: ["df"] },

  validate(ctx: PipeContext): void {
    if (ctx.df === null) {
      throw new Error("ScanStage: no df in context");
    }
  },

  async run(ctx: PipeContext): Promise<StageResult> {
    const rows = ctx.df ?? [];
    const data = new TabularData(rows);
    const stageCfg = ctx.stageConfig;
    const opts = stageCfg && Object.keys(stageCfg).length > 0 ? stageCfg : undefined;
    const result = scanData(data, opts as Parameters<typeof scanData>[1]);

    const findings: NormalizedFinding[] = result.findings.map(normalizeFinding);
    const columnProfiles = result.profile.columns;

    ctx.artifacts["findings"] = findings;
    ctx.artifacts["profile"] = result.profile;

    // Build column contexts for downstream stages (best-effort enrichment).
    try {
      const profileLikes: ColumnProfileLike[] = columnProfiles.map(toColumnProfileLike);
      const findingLikes: FindingLike[] = findings.map((f) => ({
        column: f.column,
        check: f.check,
        message: f.message,
      }));
      ctx.artifacts["column_contexts"] = buildContextsFromCheck(findingLikes, profileLikes);
    } catch {
      ctx.artifacts["column_contexts"] = [];
    }

    return { status: StageStatus.SUCCESS };
  },

  rollback: null,
};

File: packages/python/goldenpipe/goldenpipe/adapters/check.py
"""GoldenCheck adapter -- scans the in-memory frame (scan_dataframe), falling
back to scan_file() only when no frame is available."""
from __future__ import annotations

import logging

from goldenpipe.models.context import PipeContext, StageResult, StageStatus
from goldenpipe.models.stage import StageInfo

logger = logging.getLogger(__name__)

try:
    from goldencheck import scan_dataframe as _scan_df
    from goldencheck import scan_file as _scan
    HAS_CHECK = True
except ImportError:
    HAS_CHECK = False
    _scan = None
    _scan_df = None


class ScanStage:
    info = StageInfo(name="goldencheck.scan", produces=["findings", "profile"], consumes=["df"])
    rollback = None

    def validate(self, ctx: PipeContext) -> None:
        if not HAS_CHECK:
            raise RuntimeError("GoldenCheck not installed. Run: pip install goldenpipe[check]")

    def run(self, ctx: PipeContext) -> StageResult:
        source = ctx.metadata.get("source", "")
        stage_cfg = ctx.stage_config
        if ctx.df is not None:
            # Scan the frame the pipeline already loaded -- no redundant disk
            # re-read, and it works for DataFrame/DuckDB sources whose `source`
            # string ("<DataFrame>"/"duckdb:...") is not a readable path.
            # `file_path` is cosmetic (populates DatasetProfile.file_path).
            if stage_cfg:
                logger.info("Passing stage config to GoldenCheck scan_dataframe")
                result = _scan_df(ctx.df, file_path=source, **stage_cfg)
            else:
                result = _scan_df(ctx.df, file_path=source)
        else:
            # Defensive fallback: no in-memory frame. A local stage normally has
            # ctx.df materialized by the Runner before it runs, so this path is
            # rare; keep the file scan for it.
            logger.info("ScanStage: no ctx.df; falling back to scan_file(%s)", source)
            result = _scan(source, **stage_cfg) if stage_cfg else _scan(source)

        # scan_dataframe/scan_file both return a (findings, profile) tuple
        if isinstance(result, tuple) and len(result) >= 2:
            raw_findings, profile = result[0], result[1]
        else:
            raw_findings = result.findings if hasattr(result, "findings") else []
            profile = None
            logger.warning(
                "ScanStage: scan returned %s (expected tuple). "
                "Profile will be None — column context pipeline may not produce contexts.",
                type(result).__name__,
            )

        if not isinstance(raw_findings, (list, tuple)):
            logger.warning("ScanStage: raw_findings is %s, treating as empty", type(raw_findings).__name__)
            raw_findings = []

        findings = []
        for f in raw_findings:
            if isinstance(f, dict):
                findings.append(f)
            else:
                findings.append({
                    "severity": getattr(f, "severity", "info"),
                    "check": getattr(f, "check", "unknown"),
                    "column": getattr(f, "column", ""),
                    "message": getattr(f, "message", ""),
                })

        ctx.artifacts["findings"] = findings
        ctx.artifacts["profile"] = profile

        # Build column contexts for downstream stages (best-effort enrichment)
        try:
            from goldenpipe.models.column_context import build_contexts_from_check
            ctx.artifacts["column_contexts"] = build_contexts_from_check(raw_findings, profile)
        except Exception:
            logger.exception("Failed to build column contexts; downstream stages will auto-configure")
            ctx.artifacts["column_contexts"] = []

        # Advisory repair-plan (never mutates the stage list; failures are non-fatal)
        try:
            from goldenpipe.repair_host import attach_repair_plan
            _findings = ctx.artifacts.get("findings", [])
            _contexts = ctx.artifacts.get("column_contexts", [])
            if ctx.df is not None and _findings and _contexts:
                attach_repair_plan(ctx, _findings, _contexts, ctx.df)
        except Exception:
            logger.exception("repair-plan attach failed; advisory artifact skipped")

        return StageResult(status=StageStatus.SUCCESS)
